httpcore5/src/main/java/org/apache/hc/core5/http/impl/io/DefaultBHttpClientConnection.java

```diff
@@ -154,64 +197,81 @@ public void sendRequestEntity(final ClassicHttpRequest request) throws HttpExcep
             throw new LengthRequiredException();
         }
 
+
         try (final OutputStream outStream = createContentOutputStream(
-                len, this.outbuffer, new OutputStream() {
+                len,
+                this.outbuffer,
+                createResponseOutOfOrderCheckingOutputStream(socketHolder, request),
+                entity.getTrailers())) {
+            entity.writeTo(outStream);
+        } catch (final ResponseOutOfOrderException ex) {
+            if (len > 0) {
+                this.consistent = false;
+            }
+        }
+    }
 
-                    final boolean ssl = socketHolder.getSocket() instanceof SSLSocket;
-                    final InputStream socketInputStream = socketHolder.getInputStream();
-                    final OutputStream socketOutputStream = socketHolder.getOutputStream();
+    private OutputStream createResponseOutOfOrderCheckingOutputStream(
+            final SocketHolder socketHolder,
+            final ClassicHttpRequest request) throws IOException {
+        final OutputStream socketOutputStream = socketHolder.getOutputStream();
+        final int responseOutOfOrderIntervalBytes =
+                responseOutOfOrderStrategy.determineResponseCheckIntervalBytes(request);
+        if (responseOutOfOrderIntervalBytes <= 0) {
+            // Response out of order checks are disabled
+            return socketOutputStream;
+        }
 
-                    long totalBytes = 0;
-                    long chunks = -1;
+        return new OutputStream() {
 
-                    void checkForEarlyResponse() throws IOException {
-                        final long n = totalBytes / (8 * 1024);
-                        if (n > chunks) {
-                            chunks = n;
-                            if (ssl ? isDataAvailable(Timeout.ONE_MILLISECOND) : (socketInputStream.available() > 0)) {
-                                throw new ResponseOutOfOrderException();
-                            }
-                        }
-                    }
+            final InputStream socketInputStream = socketHolder.getInputStream();
+            final boolean requiresRead = !(socketInputStream instanceof FileInputStream);
```

Comment:
I do not really under stand this. How does the Socket API refactoring affect this?! How can a socket input stream be a `FileInputStream`. Can we really move this change bit to a seperate PR please? I'd like to focus on the new interface first.

Reply:
Sorry Michael, I posted on the jira ticket that I was pulling this part into a separate PR with its own jira ticket:
https://github.com/apache/httpcomponents-core/pull/207
https://issues.apache.org/jira/browse/HTTPCORE-640

Happy to discuss there.

Reply:
Yes, please undo this change here. I will go over to #207 when this one is merged.